Interpreter EH support in the runtime

janvorli opened the pull request:
This change adds support for exception handling for cases when interpreter frames either throw an exception or the exception is propagated through them. It doesn't add all the exception handling support to the interpreter itself, there is a follow up PR that contains that code.

I had to reorder the code around handling thread abort when resuming after catch for the case when the resume occurs in the interpreted code. The resuming goes to the native context of the `InterpExecMethod` in that case, so we need to update the current context in the `REGDISPLAY` to point there before we extract the `dwResumePC`. The `InterpreterFrame` stores copies of the registers that we reuse for interpreted frames stack walking and restores them before resuming there and at the same time stores the interpreter PC and SP for the resuming is stored in the `InterpreterFrame`

janvorli committed:
Interpreter EH support in the runtime

This change adds support for exception handling for cases when
interpreter frames participate in the process. That means when an
exception is thrown from an interpreter frame or when it is propagated
over interpreter frames. This doesn't add all the exception handling
support to the interpreter itself, there is a follow up PR that contains
that code.

janvorli committed:
Fix arm64 build

BrzVlad commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -1074,6 +1078,53 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
+            case INTOP_THROW:
+            {
+                CONTEXT exceptionContext;
+                ClrCaptureContext(&exceptionContext);
```
I thought we wouldn't be using register context based resume into the interpreter since this approach can't work on WASM. I think it would be better if we could use C++ try/catch to resume here instead.

janvorli replied:
Exception for resuming after catch would not work for non-WASM targets. On unix, we cannot propagate C++ exceptions through managed code and while on Windows we possibly could, it would get rid of all the perf benefits gained by the new exception handling. 
For WASM, I am planning to use setjmp / longjmp.

BrzVlad replied:
After some offline discussion with @janvorli  it would seem that c++ throw can't unwind over managed jit/aot frames since they don't have the unwind info. Apparently wasm does support setjmp/longjmp (which probably is implemented behind the scenes with try/catch) so we should be able to use a similar approach on wasm as well.

janvorli replied:
But what would you do with managed frames that we need to skip during the operation? We would need to skip them, resume at the other end, throw some special propagation exception, catch it at the other boundary and rethrow it as native exception. And there can be possibly many transitions between managed and native on the stack between the point where the catch funclet has returned and the location we need to resume to. It would just add a lot of overhead instead of single context restore on non-wasm. I don't see a benefit in doing it that way.

BrzVlad replied:
Note that I'm not sure a single context restore will be enough. I think we should unwind the native code from interpreter execution gracefully, either via return or c++ throw. So we would have a resume at the boundary of each interpreter block. Seems like forcefully resuming to the destination might break things.

jkotas replied:
> We would need to skip them, resume at the other end, throw some special propagation exception, catch it at the other boundary and rethrow it as native exception.

Right, something alone these lines. Performance of exception is not something we need to worry about with interpter.

> I don't see a benefit in doing it that way.

The benefit that I see is more portable codebase. I think about interpreter as a component optimized for portability.

janvorli replied:
The only potential issue would be that destructors for stack allocated objects in the InterpExecMethod won't be called. So we just need to make sure we don't use such objects there. Please note that we skip some native frames in coreclr EH too, e.g. when you throw a software exception, there is IL_Throw native method that calls DispatchManagedException native method which calls the managed Ex.RhThrowEx method. The IL_Throw and DispatchManagedException are always skipped during EH resuming.

janvorli replied:
@jkotas, if we can use longjmp on WASM, how does it make it less portable? Also, for the non-wasm platforms, throwing native exception at native to managed boundary has the same level of portability as restoring the context inside the InterpExecMethod.

jkotas replied:
> I don't see a problem with such an optimization in the context of the usage in the interpreter,

I meant that the [outliner](https://github.com/search?q=org%3Allvm%20Outliner&type=code) optimization can transform the code to look like this:

```
NOINLINE void Outlined(CONTEXT* p)
{
    ClrCaptureContext(p);
}

...
      CONTEXT exceptionContext;
      Outlined(&exceptionContext);
...
```

I think that this type of optimization would break this. The context captured inside `Outlined` is not resumable since the stackframe where it was captured is gone by the time you try to resume it.

The transformation that I have showed above is just to demonstrate the type of transformation done by the outliner. It would not be a profitable optimization as shown. The situation where it would cause a real problem would be much more complicated.

> the setjmp / longjmp here, because the compiler should not play any tricks around it.

Yes, use of setjmp / longjmp has to be treated as a C/C++ compiler intrinsic that disables incompatible optimizations.

> Use setjmp to store the resume location and longjmp to resume there

My preference would be to use setjmp/longjmp since it will be the same on all platforms that we care about, and we won't need to worry about C/C++ compiler introducing inscrutable bugs. The handling of setjmp exception on Windows does not need to be perfect. I would be fine with a shortcut that is active only when the interpreter is on.

BrzVlad replied:
For reference, on mono we used low level context resuming tricks similar to this (resuming to a label instead). It felt ugly and we ran into issues with the compiler breaking this so we ended up always disabling optimizations for the methods where we do this. For example https://github.com/dotnet/runtime/blob/main/src/mono/mono/mini/interp/interp.c#L2459

janvorli committed:
Fix few issues

janvorli committed:
Fix non-windows amd64 builds

ifdef-out setting m_SSP in the InterpreterFrame

janvorli committed:
Fix x86, arm and arm64 build break

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -3929,6 +3935,21 @@ extern "C" CLR_BOOL QCALLTYPE SfiNext(StackFrameIterator* pThis, uint* uExCollid
         goto Exit;
     }
 
+    #ifdef FEATURE_INTERPRETER
```
```suggestion
#ifdef FEATURE_INTERPRETER
```

janvorli committed:
Fix incorrect assert

We cannot try to get the code manager from a crawl frame when it is not
on a frameless frame.

BrzVlad commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -1074,6 +1078,53 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
+
+                TADDR resumeSP;
+                TADDR resumeIP;
+                pInterpreterFrame->GetAndClearResumeContext(&resumeSP, &resumeIP);
```
Could we have all this throwing logic extracted in a separate method (like mono's `interp_throw`). There will be tons of places where we will throw exceptions as part of different opcodes, like null checks, ovf checks etc.

jkotas replied:
I would expect that exceptions from things like null checks are going to be thrown using `COMPlusThrow`, e.g. `COMPlusThrow(kNullReferenceException)`, just like exceptions are thrown everywhere else in the VM. C++ exception is then going to caught by try/catch that's wraps the interpreter execution loop and sent for managed exception processing from there. @janvorli Does this sound right?

BrzVlad replied:
Those scenarios are logically identical to `INTOP_THROW` so I would say they should be handled the same way. You could implement `CEE_LDIND` for example either as a single `INTOP_LDIND` opcode that does the null check as part of the opcode and throws the exception or as multiple granular opcodes: `bne; throw; ldind_unsafe;`. Also I'm not sure why we would want to throw an additional c++ exception when we could do the managed exception dispatch directly.

Also resuming the context from the c++ catch handler back inside the try (since execution might resume in those frames) seems dubious. It also seems unclear whether wasm would work in this configuration, given longjmp is probably implemented via an exception throw.

jkotas replied:
The interpret needs to able to deal with C++ exceptions thrown in the rest of the VM.. For example, https://github.com/dotnet/runtime/pull/114529/files#diff-13492d6b1897666f5322f30e5f2ae4a9b35f084ce84d7f78827541070cc57329R990 can throw C++ exception that needs to be converted to managed exception and handled as managed exception. How is it going to be done? Is there a viable alternative to resuming from the C++ catch handler?

I think that the exceptions like NullReferenceException thrown by the interpret are more like C++ exceptions thrown in the rest of the VM, so we may want to use the same mechanism for them. It follows what we do with regular JIT. For example, calling an interface method on null pointer can end up here: https://github.com/dotnet/runtime/blob/b951b3d262f974e528e60e7283aec36d7e0183b0/src/coreclr/vm/virtualcallstub.cpp#L1715 . We throw C++ exception that gets converted to managed NullReferenceException like any other C++ exception thrown by the VM. In theory, we can avoid the C++ exception and throw managed NullReferenceException directly, but it would be more complicated for no good reason.

BrzVlad replied:
The way I imagined solving this problem would be to wrap calls into the runtime that might throw with a try/catch. For example:
```
try {
    pMD = pMD->GetMethodDescOfVirtualizedCode(pThisArg, pMD->GetMethodTable());
} catch (ex) {
    ex_managed = convert_to_managed (ex);
}
if (ex_managed) {
    ThrowManagedEx(ex_managed);
     // resume to the right interp frame here
}
```
And for all exceptions produced by interpreter code (that we know exactly), we would directly use `ThrowManagedEx`. Seems like this approach would be the simplest, I don't expect a ton of callsites where we would need to do this.

cshung replied:
With respect to the current implementation (i.e. not the discussion with C++ try/catch), I have found a bug with that.

By the time we resume into the interpreter to execute the logic after the catch block, the right context to restore into is not the same as the context capture during throw.

In this example, things will fail.

```c#
try
{
  try
  {
    throw null;
  }
  catch
  {
    throw null;
  }
}
catch
{
}
// During the second resume, I am about to resume to here ...
```

During the second throw, the caller has the following stack (callee on top, caller at the bottom convention)


```
DispatchException
InterpExecMethod calling funclet    // The second throw captured this context
DispatchException
InterpExecMethod calling function
```

When the second catch resumes, we really shouldn't have the `CallFunclet` on the native stack.

jkotas commented:
> The way I imagined solving this problem would be to wrap calls into the runtime that might throw with a try/catch.

it would be useful to demonstrate that this works on one callsite as part of this PR.

cshung commented:
> it would be useful to demonstrate that this works on one callsite as part of this PR.

I have been working with @janvorli on exception handling for the interpreter. On my side, building on top of this change, we know that resuming from `InterpreterCodeManager::CallFunclet` works in various cases, at least on Windows where I routinely work with.

https://github.com/dotnet/runtime/compare/main...cshung:runtime:public/interpreter-exception

> Note, it is still a work in progress. There are known scenario that doesn't work yet.

janvorli commented:
> it would be useful to demonstrate that this works on one callsite as part of this PR.

As Andrew said, without his changes, there is not much we can demonstrate here. With his changes, we can do that.

janvorli committed:
Resuming after catch via native exception

This commit moves the resuming after catch to using native exception
handling instead of fragile context capturing, which was not correct
anyways.

It also adds handling of exceptions comming out of native runtime
methods called from the interpreter.

janvorli commented:
I've just added a commit that moves the resuming after catch to a helper native exception propagation plan. When resuming in an interpreted frame, we virtually unwind from the current context upto the first native frame that is in the same block of native frames as the InterpExecMethod of the target frame and then throw the special exception from there.
I've also added cleanup of the localloc stuff during EH.
Besides that, I've added handling of exceptions stemming from native runtime calls from the interpreter.
I had to update the GCFrame so that when it is popped from the chain, it destructor doesn't do anything. To do that, I've changed its "next" pointer value indicating the end of list to the same plan as the Frame derived frames have. That means that NULL indicates the frame is not on the list at all.

jkotas commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -934,253 +942,305 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
-                    pInterpreterFrame->SetTopInterpMethodContextFrame(pFrame);
-                    GCX_COOP();
-                    GCHeapUtilities::GetGCHeap()->GarbageCollect(-1, false, 0x00000002);
+                    CONTEXT exceptionContext;
```
`CONTEXT` is a big structure. Can the CONTEXT variable and the rest of this block be in a noinline helper method so that it does not affect the rest of the method?

janvorli replied:
This is a leftover from the previous way that was using it for resuming here. I am going to remove it and call `DispatchManagedException` without the context argument. That one captures the context inside.

jkotas commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -934,253 +942,305 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
-                    GCX_COOP();
-                    GCHeapUtilities::GetGCHeap()->GarbageCollect(-1, false, 0x00000002);
+                    CONTEXT exceptionContext;
+                    ClrCaptureContext(&exceptionContext);
```
How much do we actually use in this context? I assume that it is just a tiny part. Would it make sense to capture just what's actually used?

jkotas replied:
Ah ok, I keep forgetting about the floating-point registers special case.

jkotas replied:
> I guess we could change it to capture just the non-volatile floating point registers and PC/FP/SP (for all the EH even without interpreter).

We need to deal with the non-volatile floating-point registers on Windows x64 only. Is that right? It may make sense to save everything on Windows x64 only.

In any case, this is not something to address as part of this change.

janvorli replied:
It is also needed for ARM64 on all platforms, the D8..D15 are non-volatile there. Similar for RISCV64 and LOONGARCH64 that also have non-volatile floating point registers. And ARM32 too.
Linux x64 is the only one that doesn't have any non-volatile floating point registers.

jkotas commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -934,253 +942,305 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
+            if (pFrame == NULL)
+            {
+                // Continue propagating the resume exception, the resume frame is not in this block of interpreted frames.
+                throw;
```
Would it be better to just `return` here so that we do not need `INSTALL_RESUME_AFTER_CATCH_HANDLER`? Or do you expect that `INSTALL_RESUME_AFTER_CATCH_HANDLER` is going to be used in more places?

janvorli replied:
This should actually not be here for non-wasm. Since we always resume at the correct InterpExecMethod, we should always find the frame in the loop above. Also, the INSTALL_RESUME_AFTER_CATCH_HANDLER is not needed with this change. I am making the changes in a separate checkout where I test it with Andrew's changes and I've accidentally moved some of the changes that I didn't want to. 

The side note is that I have originally made the change in a way that propagates the exception through all sequences of native frames on the stack, only skipping the managed frames. So if there were multiple sequences with InterpExecMethod, each would catch the exception here, but the intermediate ones would not find the frame and rethrow. The INSTALL_RESUME_AFTER_CATCH_HANDLER is then also placed in the `CallCatchFunclet` and `CallFinallyFunclet` and when the `InterpExecMethod` is called to execute a funclet, the resume that passes through that funclet would reach the resume after catch handler in one of the `CallXXXFunclet`, skip the managed frames of the EH code and then get rethrown in the `CallDescrWorker` that called the managed EH stuff.
This is very close to how we are planning to resume after exception on WASM, except that we would not have the resume after catch handler and the exception would flow through the AOTed frames.
I have originally wanted to use this mechanism, I have it actually working, but then I have decided to use the optimized resuming that resumes directly in the right `InterpExecMethod`. The reason is actually not because I would think that the optimized resuming necessarily performs better, because there would be the same number of unwinds going on. But I didn't like the fact that if we resume in JIT/AOTed frame, the native frames sequences are handled differently. We just skip them all in that case and resume in the JIT/AOTed code directly. While I could possibly unify it by changing it so that when the resuming would pass over native frames sequences with `InterpExecMethod`, we would behave the same way as in the case when we resume in an interpreted frame, this would add an unnecessary overhead of the extra stack walks / exception handling.

jkotas commented:
> I had to update the GCFrame so that when it is popped from the chain, it destructor doesn't do anything.

Why do we need that? I would expect that the core interpreter won't need to call GC protect, and any (throwing) calls to the VM would need to be behind unwind-and-continue wrapper (like Vlad suggested in https://github.com/dotnet/runtime/pull/114649#discussion_r2049091141) that that will to natural C++ uwnwind.

jkotas commented on src/coreclr/vm/stackwalk.cpp:
```diff
@@ -2867,12 +2867,9 @@ void StackFrameIterator::ProcessCurrentFrame(void)
-                    SetSP(pRD->pCurrentContext, dac_cast<TADDR>(pTOSInterpMethodContextFrame));
-                    SetFP(pRD->pCurrentContext, (TADDR)pTOSInterpMethodContextFrame->pStack);
-                    pRD->pCurrentContext->ContextFlags = CONTEXT_CONTROL;
+                    ((PTR_InterpreterFrame)m_crawl.pFrame)->SetContextToInterpMethodContextFrame(pRD->pCurrentContext);
```
Is this going to work for nested stackwalks in the presence of nested exceptions? In other words, are we always going to stash the exact same m_interpExecMethodIP/SP values in the InterpreterFrame?

It feels like this wants to be a state on the stackwalker instead.

janvorli replied:
The `SetContextToInterpMethodContextFrame` gets the topmost context of the interpreted frames in the `InterpExecMethod` belonging to the `InterpreterFrame`. These IP/SP are not stashed in the frame.

At the point where the stack walker reaches the `InterpreterFrame`, it needs to enumerate all the underlying interpreted frames and this just sets the pRD->pCurrentContext to the topmost one of them for that `InterpExecMethod`

jkotas replied:
Would it make sense to add some asserts to guarantee that nested or parallel stackwalks are not trying to store conflicting values into `m_interpExec...` fields?

BrzVlad commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -49,263 +53,267 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
+        try
         {
+            INSTALL_MANAGED_EXCEPTION_DISPATCHER;
+            INSTALL_UNWIND_AND_CONTINUE_HANDLER;
```
I don't think this code is meant to be run for each opcode executed, we could at least extract it out of the loop ? Any reason why this couldn't be a simple try/catch in the future, so we don't have any overhead at all on the non-exceptional path ?

janvorli commented:
> Why do we need that? I would expect that the core interpreter won't need to call GC protect, and any (throwing) calls to the VM would need to be behind unwind-and-continue wrapper (like Vlad suggested in [#114649 (comment)](https://github.com/dotnet/runtime/pull/114649#discussion_r2049091141)) that that will do natural C++ uwnwind.

@jkotas this is related to the lengthy comment above related to your question on the `INSTALL_RESUME_AFTER_CATCH_HANDLER `. This state of the change doesn't need it, it was another thing that I've moved accidentally to this change, as it was needed for the case when we do the "full" propagation over all intermediate native frames. However, it will be needed for WASM, because in that case, the exception will go through the `DispatchManagedException` and that one has GC_PROTECT stuff in it around the call to the managed EH code.

janvorli committed:
Remove some unneeded stuff and fix Unix build

jkotas commented:
> However, it will be needed for WASM, because in that case, the exception will go through the DispatchManagedException and that one has GC_PROTECT stuff in it around the call to the managed EH code.

I would expect WASM exception handling to call the C++ destructor naturally whenever it unwinds a C++ frame. Is it not going to be the case?

jkotas commented on src/coreclr/vm/interpexec.cpp:
```diff
@@ -47,265 +51,269 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
-
-        switch (*ip)
+        INSTALL_MANAGED_EXCEPTION_DISPATCHER;
+        INSTALL_UNWIND_AND_CONTINUE_HANDLER;
```
Is `INSTALL_UNWIND_AND_CONTINUE_HANDLER` going to break exception filter semantics when we have mix of interpreted and JIIT/AOT frames on the stack?

jkotas replied:
I guess it works fine with filter. I do not understand why we need `INSTALL_UNWIND_AND_CONTINUE_HANDLER` here. What is going to break if `INSTALL_UNWIND_AND_CONTINUE_HANDLER` is deleted here?

janvorli replied:
That's for handling exceptions coming out of calls to native runtime. The INSTALL_MANAGED_EXCEPTION_DISPATCHER does nothing for Windows. Without it, those exceptions would just flow through.

janvorli commented:
> I would expect WASM exception handling to call the C++ destructor naturally whenever it unwinds a C++ frame. Is it not going to be the case?

Yes and that's why that change was added. Without it, the destructor of the GCFrame would try to pop it out of the frame list and it would assert, because it would find the frame is not the first on the list. The other way to solve that would be to not to pop the GCFrames upfront when resuming after catch and let the destructor pop them. However, that would make it different from when those frames are popped on WASM and non-WASM and I'd prefer this to be uniform.

janvorli commented:
I am not sure I understand. This change ensures that it works the same way with and without the interpreter. The frames with the GC protects are virtually dead.

jkotas commented:
> I am not sure I understand. This change ensures that it works the same way with and without the interpreter. The frames with the GC protects are virtually dead.

Ok, I have missed that we have special handling for GCFrames in PopExplicitFrames. My mental model for `GCFrame`s was that they are ordinary C++ holders with destructors called by C++ runtime, but it is not actually the case.

Could you please add a comment to `PopExplicitFrames` about the GCFrames that require us to do the pop them explicitly? I assume that there are a very few of those, most of them are on the ordinary C++ holder plan.

jkotas commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -3002,23 +3002,70 @@ extern "C" void QCALLTYPE AppendExceptionStackFrame(QCall::ObjectHandleOnStack e
+// The SECOND_ARG_REG is defined only for Windows, it is used to handle longjmp propagation over managed frames
+#ifdef HOST_AMD64
+#ifdef UNIX_AMD64_ABI
+#define FIRST_ARG_REG Rdi
```
Can this be in cgencpu.h and de-duped with the `GetFirstArgReg` that you have added in this change?

janvorli commented:
> Make sense. This means that the interpreter should be able to use regular COMPlusThrow to throw exceptions like I originally assumed in https://github.com/dotnet/runtime/pull/114649#discussion_r2049008891

Yes, I've verified that it works already.

janvorli commented on src/coreclr/vm/frames.cpp:
```diff
@@ -1136,19 +1136,25 @@ GCFrame::~GCFrame()
     {
-        m_pCurThread->DisablePreemptiveGC();
-    }
+        // This is a GCFrame that was not popped.  This is a problem.
+        // We should have popped it before we destruct
```
Ugh. I didn't write this comment except for the first sentence. I haven't noticed copilot has added the nonsense after.

janvorli commented:
> Would it make sense to add some asserts to guarantee that nested or parallel stackwalks are not trying to store conflicting values into m_interpExec... fields?

I am actually going to do that differently, storing those in the stack walker state instead. The storage in the frame is a remainder of the first version when I was using that during restoring context after catch and I think that keeping it in that state didn't work for some reason.

janvorli committed:
Move the saved InterpExecMethod context regs

* Move the saved registers to the StackFrameIterator
* Fix MUSL build
* Add {Get/Set}{First/Second}ArgumentRegister to all architectures and use
  those in ExecuteFunctionBelowContext.

janvorli committed:
Few fixes

janvorli committed:
Fix builds with disabled interpreter

janvorli committed:
One more MUSL build fix

janvorli committed:
Implement proper GetFuncletStartAddress for interpreter

jkotas commented on src/coreclr/vm/codeman.cpp:
```diff
@@ -4339,8 +4339,25 @@ BOOL InterpreterJitManager::JitCodeToMethodInfo(
+        EE_ILEXCEPTION_CLAUSE EHClause;
+        pJitMan->GetNextEHClause(&EnumState, &EHClause);
+
+        if (EHClause.HandlerStartPC <= relOffset && relOffset < EHClause.HandlerEndPC)
```
Does this work for filters?

jkotas replied:
I am not sure whether you wanted to push this into this PR since there is no matching logic in the JIT, so it is hard to tell whether it is correct.

janvorli replied:
Yes, it works for filters too. The interpreter compiler sets the `HandlerEndPC` even for filters.

> I am not sure whether you wanted to push this into this PR since there is no matching logic in the JIT, so it is hard to tell whether it is correct.

Makes sense, let's make it part of Andrew's part of the change instead.

janvorli replied:
Hmm, actually, even JIT sets the `HandlerEndPC` for filters

jkotas replied:
The problem that I see is that the filter funclet is a separate funclet that is not included in the (HandlerStartPC, HandledEndPC) range, with the JIT at least. I assume that you want to use the same encoding rules for interpreter as well.

I have done a quick test.
```
try
{
    Console.WriteLine("Try");
}
catch (Exception e) when (args.Length > 10)
{
    Console.WriteLine("Catch");
}
```
Produces 
```
<img width="155" alt="image" src="https://github.com/user-attachments/assets/820b6b67-e846-4cb8-bb17-f780d911ee49" />
```
Notice that FilterOffset is not included (HandlerStartPC, HandledEndPC).

janvorli replied:
Ah, I am sorry, you are right, the ranges I am scanning contain the catch handler for the filter, not the filter itself. That's the same thing for the interpreter then.

janvorli replied:
We will need to figure out some other mechanism for getting the end address of filter funclets. The JIT uses unwind info stuff, which we don't have.

jkotas replied:
You can add a rule that the filter funclet must immediately precede the handler. (This rule would apply to interpreter JIT only.)

janvorli replied:
Thank you, that sounds like a reasonable solution.

jkotas commented:
The failures look related to the change.

janvorli commented:
> The failures look related to the change.

These were interesting. I was missing FEATURE_INTERPRETER definition for the managed code. The offset verification was passing, because that's done in native build, but the managed code was using wrong offsets. I'll push a fix in a minute.

janvorli committed:
Fix problem with missing managed FEATURE_INTERPRETER define

janvorli committed:
Revert "Implement proper GetFuncletStartAddress for interpreter"

This reverts commit d967861d818e98af2d5eac6c6dee5492319c4353.
